File: src/main/java/aoc2015/day03/Day03.java
package aoc2015.day03;

import aoc2015.Day2015;
import util.grid.Direction;
import util.grid.Point;

import java.util.HashSet;
import java.util.Set;

class Day03 extends Day2015 {

    Day03() {
        super(3);
    }

    @Override
    public Integer part1(String input) {
        Set<Point> visited = new HashSet<>();

        Point pos = new Point(0, 0);
        visited.add(pos);
        for (char c : input.toCharArray()) {
            pos = pos.move(Direction.fromChar(c));
            visited.add(pos);
        }

        return visited.size();
    }

    @Override
    public Integer part2(String input) {
        Set<Point> visited = new HashSet<>();

        Point santa = new Point(0, 0);
        Point robot = new Point(0, 0);
        visited.add(santa);

        for (int i = 0; i < input.length(); i++) {
            char c = input.charAt(i);
            if (i % 2 == 0) {
                santa = santa.move(Direction.fromChar(c));
                visited.add(santa);
            } else {
                robot = robot.move(Direction.fromChar(c));
                visited.add(robot);
            }
        }

        return visited.size();
    }

    public static void main(String[] args) {
        new Day03().main();
    }
}

File: src/main/java/aoc2020/day03/Day03.kt
package aoc2020.day03

import aoc2020.AbstractDay2020
import util.grid.Point
import java.util.stream.Stream

fun main() {
    Day03().executeParts()
}

class Day03 : AbstractDay2020(3) {

    override fun part1(input: String): Any {
        val areaMap = parse(input)

        return descend(areaMap, Slope(3, 1))
    }

    override fun part2(input: String): Any {
        val areaMap = parse(input)

        return Stream.of(
            Slope(1, 1),
            Slope(3, 1),
            Slope(5, 1),
            Slope(7, 1),
            Slope(1, 2)
        )
            .map { descend(areaMap, it) }
            .reduce(Int::times)
            .orElse(0)
    }

    private fun descend(areaMap: AreaMap, slope: Slope): Int {
        var point = Point(0, 0)
        var treeCount = 0

        while (!areaMap.hasReachedBottom(point)) {
            if (areaMap.getTile(point) == Tile.TREE) {
                treeCount++
            }
            point = slope.nextPoint(point)
        }

        return treeCount
    }

    private fun parse(input: String): AreaMap {
        val map = input.lines()
            .map { line ->
                line.toCharArray()
                    .map { Tile.from(it) }
                    .toList()
            }
            .toList()

        return AreaMap(map)
    }
}